app/lib/classification-queue.js

```diff
@@ -0,0 +1,101 @@
+import apiClient from 'panoptes-client/lib/api-client';
+import { Split } from 'seven-ten';
+
+const FAILED_CLASSIFICATION_QUEUE_NAME = 'failed-classifications';
+const MAX_RECENTS = 10;
+
+class ClassificationQueue {
+    constructor(storage, api) {
+        this.storage = storage || window.localStorage;
+        this.apiClient = api || apiClient;
+        this.recents = [];
+    }
+
+    add(classification) {
+        var queue = this._loadQueue();
+        queue.push(classification);
+
+        try {
+            this._saveQueue(queue);
+            console.info('Queued classifications:', queue.length);
+        }
+        catch(error) {
+            console.error('Failed to queue classification:', error);
+        }
+
+        this.flushToBackend();
+    }
+
+    length() {
+        return this._loadQueue().length
+    }
+
+    addRecent(classification) {
+        if (this.recents.length > MAX_RECENTS) {
+            let droppedClassification = this.recents.pop();
+            droppedClassification.destroy();
+        }
+
+        this.recents.unshift(classification);
+    }
+
+    flushToBackend() {
+        var queue = this._loadQueue();
+
+        if (queue.length > 0) {
+            console.log('Saving queued classifications:', queue.length);
+            for (let classificationData of queue) {
+                this.apiClient.type('classifications').create(classificationData)
+                    .save()
+                    .then((actualClassification) => {
+                        console.log('Saved classification', actualClassification.id);
+                        Split.classificationCreated(actualClassification); // Metric log needs classification id
+                        this.addRecent(actualClassification);
+                        let indexInQueue = queue.indexOf(classificationData);
+                        queue.splice(indexInQueue, 1);
+                        try {
```

Comment:
I'm not sure why this is using `try…catch` instead of handling errors when the promise rejects.

Reply:
nvm. This is catching local storage errors, like 'out of space'.